File: kid-ios/MyWorld/Views/ExpressiveNamingView.swift
import SwiftUI

/// PRD §5 "Expressive naming" mode: the image is shown alone — no audio
/// prompt, no choice tiles — and the child speaks (or gestures, or selects an
/// object) the answer unaided. The facilitator's phone marks the response via
/// the existing `/api/live` `mark` command. This mode measures EXPRESSIVE
/// production, which carries more evidentiary weight than receptive matching
/// because the answer isn't on-screen to choose from.
///
/// UX:
///   - Big target image centered. No label, no description.
///   - Long-hold ✕ to exit (the universal parent-only gesture).
///   - A short tap on the image SKIPS the slide (recorded as a fail) and
///     advances, so the child can self-drive when no facilitator is present.
///   - Facilitator `mark { method }` → pass with that method, advance.
///   - Facilitator `skip` / `next` → fail, advance.
///   - Facilitator `end` → finish (with end_reason='facilitator_stop').
///   - Time limit → finish (end_reason='timeout').
///   - At session end the same `/api/game-log` payload as matching is shipped,
///     with `mode='expressive_naming'`.
struct ExpressiveNamingView: View {
    let session: GameController.Session
    let onExit: () -> Void

    @Environment(BoardStore.self) private var board
    @Environment(LiveSession.self) private var live
    @Environment(GameController.self) private var game
    @Environment(AuthManager.self) private var auth

    @State private var targets: [Tile] = []
    @State private var index = 0
    @State private var image: UIImage?
    @State private var locked = false
    @State private var celebrating = false
    @State private var finished = false
    @State private var lastHandledCmdSeq = 0
    @State private var limitTask: Task<Void, Never>?

    // Same per-attempt log shape as MatchingView so the server sees a uniform
    // payload regardless of which game mode produced it (PRD §3 honest scoring).
    @State private var startedAt: Date = .init()
    @State private var correctCount = 0
    @State private var loggedAttempts: [MatchingView.LoggedAttempt] = []
    @State private var pendingMarkMethod: String?

    private var target: Tile? { targets.indices.contains(index) ? targets[index] : nil }

    var body: some View {
        ZStack {
            Color(hex: "#fff7fb").ignoresSafeArea()

            if finished {
                finishedView
            } else if let _ = target {
                centerImage
            } else {
                Text("Nothing to practice here")
                    .font(.title2).foregroundStyle(.secondary)
            }

            LongPressExitButton.corner(
                tint: Color(hex: "#ad1457"),
                background: Color.black.opacity(0.06)
            ) { onExit() }

            ConfettiView(running: celebrating)
        }
        .contentShape(Rectangle())
        // A short tap skips this slide (recorded as fail) so the child can
        // self-drive. Long-hold of the ✕ in the corner is the exit; this is
        // an unrelated tap layer behind the X.
        .onTapGesture { skipFromTap() }
        .task { setup() }
        .onDisappear {
            limitTask?.cancel()
            GameAudio.shared.stopMusic()
        }
        .task(id: index) { await loadCurrent() }
        .onChange(of: game.inGameCommand) { _, cmd in handleCommand(cmd) }
    }

    // MARK: -- Pieces

    private var centerImage: some View {
        VStack(spacing: 16) {
            Spacer()
            if let img = image {
                Image(uiImage: img)
                    .resizable()
                    .aspectRatio(contentMode: .fit)
                    .frame(maxWidth: 820, maxHeight: 620)
                    .clipShape(RoundedRectangle(cornerRadius: 32))
                    .shadow(color: .black.opacity(0.12), radius: 28, y: 8)
                    .padding(40)
            } else {
                ProgressView().tint(Color(hex: "#ad1457"))
            }
            Spacer()
        }
    }

    private var finishedView: some View {
        VStack(spacing: 18) {
            Text("🎉").font(.system(size: 96))
            Text("Great job!")
                .font(.system(size: 52, weight: .bold, design: .rounded))
                .foregroundStyle(Color(hex: "#ad1457"))
        }
    }

    // MARK: -- Lifecycle

    private func setup() {
        let pool = board.tilesForScope(session.scope, from: session.from, to: session.to)
            .filter { $0.imageKey?.isEmpty == false }
        var picked = pool.shuffled()
        if let n = session.sample, n > 0 { picked = Array(picked.prefix(n)) }
        targets = picked
        index = 0
        correctCount = 0
        startedAt = Date()
        loggedAttempts = []
        if targets.isEmpty {
            finishGame(reason: "empty_scope")
        } else {
            GameAudio.shared.startMusic(childId: auth.childSlug)
            publishState()
        }
        startLimitTimerIfNeeded()
    }

    private func loadCurrent() async {
        image = nil
        locked = false
        pendingMarkMethod = nil
        guard let key = target?.imageKey, !key.isEmpty,
              let img = await MediaCache.shared.image(for: key) else { return }
        await MainActor.run { self.image = img }
        publishState()
    }

    private func skipFromTap() {
        // Treat a tap as "I don't know" — record a fail with method 'tap'
        // (not child-generated for spike weighting) and advance.
        guard !locked, let t = target else { return }
        locked = true
        recordAttempt(target: t, passed: false, methodOverride: "tap")
        advance(after: 0.3)
    }

    private func handleCommand(_ cmd: LiveCommand?) {
        guard let cmd, cmd.seq > lastHandledCmdSeq else { return }
        lastHandledCmdSeq = cmd.seq
        switch cmd.action {
        case "next", "skip":
            guard let t = target, !locked else { return }
            locked = true
            recordAttempt(target: t, passed: false, methodOverride: "tap")
            advance()
        case "mark":
            // PRD §4: in Expressive Naming the only "input" is the
            // facilitator's mark — there are no on-screen choices to tap.
            // Default to 'verbal' when the console doesn't specify, since
            // that's the canonical expressive-naming response.
            let method = (cmd.method?.isEmpty == false) ? cmd.method! : "verbal"
            guard let t = target, !locked else { return }
            locked = true
            correctCount += 1
            celebrate()
            // Optional reinforcement: speak the label so the child hears the
            // right word after producing it. Confirms the production.
            Task { await TilePlayer.shared.play(t) }
            recordAttempt(target: t, passed: true, methodOverride: method)
            advance(after: 0.6)
        case "end":
            finishGame(reason: "facilitator_stop")
        default:
            break
        }
        game.consumeInGameCommand()
    }

    // MARK: -- Game logic

    private func recordAttempt(target: Tile, passed: Bool, methodOverride: String? = nil) {
        let method = methodOverride ?? pendingMarkMethod ?? "tap"
        pendingMarkMethod = nil
        loggedAttempts.append(MatchingView.LoggedAttempt(
            itemId: target.id,
            label: target.label,
            category: target.taxonomySlug?.split(separator: ".").dropLast().joined(separator: ".") ?? nil,
            taxonomySlug: target.taxonomySlug,
            correct: passed,
            inputMethod: method,
            attemptsTaken: 1,            // expressive: a single facilitator decision per slide
            distractorCount: 0,          // no on-screen choices
            misses: passed ? 0 : 1,
            occurredAt: Date()
        ))
    }

    private func celebrate() {
        celebrating = true
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.4) { celebrating = false }
    }

    private func advance(after delay: TimeInterval = 0) {
        DispatchQueue.main.asyncAfter(deadline: .now() + delay) {
            if index + 1 < targets.count {
                index += 1
            } else {
                finishGame(reason: "completed")
            }
        }
    }

    private func startLimitTimerIfNeeded() {
        limitTask?.cancel()
        guard let mins = session.limitMin, mins > 0 else { return }
        let nanos = UInt64(mins * 60 * 1_000_000_000)
        limitTask = Task {
            try? await Task.sleep(nanoseconds: nanos)
            guard !Task.isCancelled else { return }
            await MainActor.run { finishGame(reason: "timeout") }
        }
    }

    private func finishGame(reason: String) {
        guard !finished else { return }
        finished = true
        live.setEnded(currentPayload())
        GameAudio.shared.playCheer(childId: auth.childSlug)
        submitGameLog(endReason: reason)
        Task {
            try? await Task.sleep(nanoseconds: 4_400_000_000)
            await MainActor.run { onExit() }
        }
    }

    // MARK: -- Live + submit

    private func publishState() {
        live.setRunning(currentPayload())
    }

    private func currentPayload() -> LivePayload {
        LivePayload(
            target: target.map { .init(label: $0.label, imageKey: $0.imageKey) },
            i: index,
            total: targets.count,
            correctCount: correctCount
        )
    }

    private func submitGameLog(endReason: String) {
        guard !loggedAttempts.isEmpty else { return }
        var skillCounts: [String: Int] = [:]
        for a in loggedAttempts {
            guard let s = a.taxonomySlug, !s.isEmpty else { continue }
            skillCounts[s, default: 0] += 1
        }
        let skillSlug = skillCounts.max(by: { $0.value < $1.value })?.key
        let payload = APIClient.GameLogPayload(
            childId: auth.childSlug,
            mode: "expressive_naming",
            category: session.scope,
            startedAt: ISO8601DateFormatter().string(from: startedAt),
            endedAt: ISO8601DateFormatter().string(from: Date()),
            itemCount: targets.count,
            slidesAttempted: loggedAttempts.count,
            correctCount: correctCount,
            scoringVersion: 2,
            endReason: endReason,
            skillSlug: skillSlug,
            attempts: loggedAttempts.map { la in
                APIClient.GameLogPayload.Attempt(
                    itemId: la.itemId,
                    label: la.label,
                    category: la.category,
                    taxonomySlug: la.taxonomySlug,
                    correct: la.correct,
                    inputMethod: la.inputMethod,
                    misses: la.misses,
                    attemptsTaken: la.attemptsTaken,
                    distractorCount: la.distractorCount,
                    childGenerated: la.inputMethod != "tap",
                    occurredAt: ISO8601DateFormatter().string(from: la.occurredAt)
                )
            }
        )
        Task.detached(priority: .utility) {
            await APIClient().submitGameLog(payload)
        }
    }
}

File: android-native/app/src/main/java/io/andrewpeterson/myworld/ui/game/ExpressiveNamingView.kt
package io.andrewpeterson.myworld.ui.game

import android.graphics.Bitmap
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.produceState
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import io.andrewpeterson.myworld.LocalAppContainer
import io.andrewpeterson.myworld.game.GameController
import io.andrewpeterson.myworld.live.LivePayload
import io.andrewpeterson.myworld.model.Tile
import io.andrewpeterson.myworld.net.GameLogPayload
import io.andrewpeterson.myworld.net.submitGameLog
import io.andrewpeterson.myworld.ui.LongPressExitButton
import io.andrewpeterson.myworld.ui.theme.Brand
import io.andrewpeterson.myworld.ui.theme.hexColor
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * Expressive naming — port of `Views/ExpressiveNamingView.swift`: the image
 * is shown ALONE (no audio prompt, no choices); the child speaks/gestures the
 * answer unaided. A short tap on the image skips (recorded fail, method tap);
 * facilitator `mark{method}` = pass; `skip`/`next` = fail; `end` stops.
 */
@Composable
fun ExpressiveNamingView(session: GameController.Session, onExit: () -> Unit) {
    val c = LocalAppContainer.current
    val scope = rememberCoroutineScope()
    val inGameCommand by c.game.inGameCommand.collectAsState()

    var targets by remember { mutableStateOf<List<Tile>>(emptyList()) }
    var index by remember { mutableIntStateOf(0) }
    var correctCount by remember { mutableIntStateOf(0) }
    var celebrating by remember { mutableStateOf(false) }
    var finished by remember { mutableStateOf(false) }
    var lastHandledCmdSeq by remember { mutableIntStateOf(0) }

    val startedAt = remember { Date() }
    val loggedAttempts = remember { mutableListOf<GameLogPayload.Attempt>() }
    val iso = remember {
        SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }
    }

    val target: Tile? = targets.getOrNull(index)

    fun payload() = LivePayload(
        target = target?.let { LivePayload.Target(it.label, it.imageKey) },
        i = index, total = targets.size, correctCount = correctCount,
    )

    fun record(t: Tile, passed: Boolean, method: String, attemptsTaken: Int = 1) {
        loggedAttempts.add(GameLogPayload.Attempt(
            itemId = t.id, label = t.label,
            category = t.taxonomySlug?.split('.')?.dropLast(1)?.joinToString(".")?.takeIf { it.isNotEmpty() },
            taxonomySlug = t.taxonomySlug, correct = passed, inputMethod = method,
            misses = if (passed) 0 else 1, attemptsTaken = maxOf(1, attemptsTaken),
            distractorCount = 0, childGenerated = method != "tap",
            occurredAt = iso.format(Date()),
        ))
    }

    fun finishGame(reason: String = "completed") {
        if (finished) return
        finished = true
        c.live.setEnded(payload())
        scope.launch { c.gameAudio.playCheer(c.auth.childSlug) }
        if (loggedAttempts.isNotEmpty()) {
            val counts = mutableMapOf<String, Int>()
            for (a in loggedAttempts) a.taxonomySlug?.takeIf { it.isNotEmpty() }?.let {
                counts[it] = (counts[it] ?: 0) + 1
            }
            val p = GameLogPayload(
                childId = c.auth.childSlug, mode = "expressive_naming", category = session.scope,
                startedAt = iso.format(startedAt), endedAt = iso.format(Date()),
                itemCount = targets.size, slidesAttempted = loggedAttempts.size,
                correctCount = correctCount, scoringVersion = 2, endReason = reason,
                skillSlug = counts.maxByOrNull { it.value }?.key,
                attempts = loggedAttempts.toList(),
            )
            scope.launch { c.api.submitGameLog(p) }
        }
        scope.launch { delay(4_400); onExit() }
    }

    fun advance(afterMs: Long = 0) {
        scope.launch {
            if (afterMs > 0) delay(afterMs)
            if (index + 1 < targets.size) { index += 1; c.live.setRunning(payload()) }
            else finishGame()
        }
    }

    LaunchedEffect(Unit) {
        var picked = c.board.tilesForScope(session.scope, session.from, session.to)
            .filter { !it.imageKey.isNullOrEmpty() }.shuffled()
        session.sample?.takeIf { it > 0 }?.let { picked = picked.take(it) }
        targets = picked
        if (targets.isEmpty()) { finishGame("empty_scope"); return@LaunchedEffect }
        c.live.setRunning(payload())
        session.limitMin?.takeIf { it > 0 }?.let { mins ->
            scope.launch { delay((mins * 60_000).toLong()); finishGame("timeout") }
        }
    }

    LaunchedEffect(inGameCommand) {
        val cmd = inGameCommand ?: return@LaunchedEffect
        if (cmd.seq <= lastHandledCmdSeq) return@LaunchedEffect
        lastHandledCmdSeq = cmd.seq
        when (cmd.action) {
            "mark" -> {
                val method = cmd.method?.takeIf { it.isNotEmpty() } ?: "verbal"
                target?.let { t ->
                    correctCount += 1
                    record(t, true, method, cmd.attemptsTaken ?: 1)
                    celebrating = true
                    scope.launch { delay(1_400); celebrating = false }
                }
                advance(600)
            }
            "next", "skip" -> { target?.let { record(it, false, "tap") }; advance() }
            "end" -> finishGame("facilitator_stop")
        }
        c.game.consumeInGameCommand()
    }

    val image by produceState<Bitmap?>(initialValue = null, target?.imageKey) {
        val key = target?.imageKey
        value = if (key.isNullOrEmpty()) null else c.media.bitmap(key, maxDim = 640)
    }

    Box(Modifier.fillMaxSize().background(hexColor("#fff7fb"))) {
        if (finished) {
            Column(Modifier.fillMaxSize(), horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.Center) {
                Text("🎉", fontSize = 96.sp)
                Text("Great job!", fontSize = 52.sp, fontWeight = FontWeight.Bold, color = Brand.pinkDeep)
            }
        } else {
            Column(
                Modifier.fillMaxSize().padding(28.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.Center,
            ) {
                val img = image
                if (img != null) {
                    Image(img.asImageBitmap(), contentDescription = null,
                        contentScale = ContentScale.Fit,
                        modifier = Modifier.widthIn(max = 720.dp).heightIn(max = 520.dp)
                            .clip(RoundedCornerShape(32.dp))
                            // Short tap = skip (fail, method tap) — the child
                            // moved on without naming it.
                            .clickable { target?.let { record(it, false, "tap") }; advance() })
                } else if (target != null) {
                    CircularProgressIndicator(color = Brand.pinkDeep)
                }
            }
        }
        LongPressExitButton(onExit = { finishGame("facilitator_stop") },
            modifier = Modifier.align(Alignment.TopEnd))
        ConfettiView(running = celebrating)
    }
}
